File: Python/07-findClosestValueInBst/python/findClosestValueInBST.py
# Find Closest Value In BST
class BST:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

def findClosestValueInBst(tree, target):
    return bst_helper(tree, target, float('inf'))

def bst_helper(tree, target, closest):
    if tree is None:
        return closest
    if abs(target - closest) > abs(target - tree.value):
        closest = tree.value
    if target < tree.value:
        return bst_helper(tree.left, target, closest)
    elif target > tree.value:
        return bst_helper(tree.right, target, closest)
    else:
        return closest

# Function to insert values into the BST
def insert_into_bst(tree, value):
    if tree is None:
        return BST(value)
    if value < tree.value:
        tree.left = insert_into_bst(tree.left, value)
    else:
        tree.right = insert_into_bst(tree.right, value)
    return tree

# Create a BST and insert values
root = BST(10)
insert_into_bst(root, 5)
insert_into_bst(root, 2)
insert_into_bst(root, 5)
insert_into_bst(root, 1)
insert_into_bst(root, 15)
insert_into_bst(root, 13)
insert_into_bst(root, 22)
insert_into_bst(root, 14)

# Call the findClosestValueInBst function
target = 12
closest_value = findClosestValueInBst(root, target)
print(f"The closest value to {target} in the BST is {closest_value}")

File: JavaScript/findClosestValueInBST/findClosestValueInBST.js
// This is the class of the input tree.
class BST {
	constructor(value) {
	  this.value = value;
	  this.left = null;
	  this.right = null;
	}
  }
  
  function findClosestValueInBst(tree, target) {
	return bstHelper(tree, target, Infinity);
  }
  
  function bstHelper(tree, target, closest) {
	if (tree === null) {
	  return closest;
	}
	if (Math.abs(target - closest) > Math.abs(target - tree.value)) {
	  closest = tree.value;
	}
	if (target < tree.value) {
	  return bstHelper(tree.left, target, closest);
	} else if (target > tree.value) {
	  return bstHelper(tree.right, target, closest);
	} else {
	  return closest;
	}
  }
  
  function insertIntoBst(tree, value) {
	if (tree === null) {
	  return new BST(value);
	}
	if (value < tree.value) {
	  tree.left = insertIntoBst(tree.left, value);
	} else {
	  tree.right = insertIntoBst(tree.right, value);
	}
	return tree;
  }
  
  // Create a BST and insert values
  let root = new BST(10);
  insertIntoBst(root, 5);
  insertIntoBst(root, 2);
  insertIntoBst(root, 5);
  insertIntoBst(root, 1);
  insertIntoBst(root, 15);
  insertIntoBst(root, 13);
  insertIntoBst(root, 22);
  insertIntoBst(root, 14);
  
  // Call the findClosestValueInBst function
  const target = 12;
  const closestValue = findClosestValueInBst(root, target);
  console.log(`The closest value to ${target} in the BST is ${closestValue}`);
  